src/backend/distributed/executor/multi_utility.c

```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
-			uint32 workerPort = placement->nodePort;
+		ExecuteCommandOnShardReplicas(applyCommand, shardId, shardConnections);
 
-			List *queryResultList = ExecuteRemoteQuery(workerName, workerPort,
-													   relationOwner, applyCommand);
-			if (queryResultList == NIL)
-			{
-				/*
-				 * If we failed on the first placement, return false. We return
-				 * here instead of exiting at the end to avoid breaking through
-				 * multiple loops.
-				 */
-				if (isFirstPlacement)
-				{
-					return false;
-				}
+		FreeStringInfo(applyCommand);
+	}
 
-				ereport(WARNING, (errmsg("could not apply command on shard "
-										 UINT64_FORMAT " on node %s:%d", shardId,
-										 workerName, workerPort),
-								  errdetail("Shard placement will be marked as "
-											"inactive.")));
+	/* check for cancellation one last time before returning */
+	CHECK_FOR_INTERRUPTS();
 
-				*failedPlacementList = lappend(*failedPlacementList, placement);
-			}
-			else
-			{
-				ereport(DEBUG2, (errmsg("applied command on shard " UINT64_FORMAT
-										" on node %s:%d", shardId, workerName,
-										workerPort)));
-			}
+	/* CloseConnections(allShardsConnectionList); */
 
-			isFirstPlacement = false;
-		}
+	return true;
+}
 
-		FreeStringInfo(applyCommand);
+
+/*
+ * OpenConnectionsToAllShardPlacement opens connections to all placements of
+ * the given shard Id Pointer List and returns the hash table containing the connections.
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+static HTAB *
+OpenConnectionsToAllShardPlacements(List *shardIdList, char *relationOwner)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
+	ListCell *shardCell = NULL;
+	MemoryContext oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+	foreach(shardCell, shardIdList)
+	{
+		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
+		uint64 shardId = (*shardIdPointer);
+
+		OpenConnectionsToShardPlacements(shardId, shardConnectionHash, relationOwner);
 	}
 
-	return true;
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
+
+	MemoryContextSwitchTo(oldContext);
+
+	if (!isXactCallbackRegistered)
+	{
+		RegisterXactCallback(CompleteShardPlacementTransactions, NULL);
+		isXactCallbackRegistered = true;
+	}
+
+	return shardConnectionHash;
+}
+
+
+/*
+ * CompleteShardPlacementTransactions commits or aborts pending shard placement
+ * transactions when the local transaction commits or aborts.
+ */
+static void
+CompleteShardPlacementTransactions(XactEvent event, void *arg)
+{
+	if (shardPlacementConnectionList == NIL)
+	{
+		/* nothing to do */
+		return;
+	}
+	else if (event == XACT_EVENT_PRE_COMMIT)
+	{
+		/*
+		 * Any failure here will cause local changes to be rolled back,
+		 * and remote changes to either roll back (1PC) or, in case of
+		 * connection or node failure, leave a prepared transaction
+		 * (2PC).
+		 */
+
+		PrepareRemoteTransactions(shardPlacementConnectionList);
```

Comment:
I had removed the multi_shard_commit_protocol reference because it seemed to me that all DDL propagations should have been done via 2PC. I put it back now.